contracts/SoulBoundToken/BasicSBT.sol

```diff
@@ -0,0 +1,215 @@
+// SPDX-License-Identifier: MIT
+// SpartanLabs Contracts (SBT)
+pragma solidity ^0.8.0;
+
+import "./utils/Ownable.sol";
+
+/**
+ * @dev Implementation of Soul Bound Token (SBT)
+ * following Vitalik's co-authored whitepaper at: https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4105763
+ *
+ * Contract provides a basic Soul Bound Token mechanism, where address can mint SBT with their private data.
+ *
+ * By default, the owner account will be the one that deploys the contract. This
+ * can later be changed with {transferOwnership}.
+ */
+contract BasicSBT is Ownable {
+    // Name for the SBT
+    string public _name;
+
+    // Symbol for the SBT
+    string public _symbol;
+
+    // Total count of SBT
+    uint256 public _totalSBT; // TODO: Should we keep track of how many SBT?
+
+    // Mapping from SBT ID to owner address
+    mapping(uint256 => address) private _owners; // TODO : Do we even need to keep track of owners?
+
+    // Mapping between address and the soul
+    mapping(address => Soul) private souls;
+
+    /**
+     * @dev Struct `Soul` contains the soulbound token information for a given address.
+     * The fields within the struct can be be edited for the usecase of the SBT.
+     *
+     * The fields of `identity` and `url` can be hashed for privacy.
+     */
+    struct Soul {
+        string identity;
+        string url;
+    }
+
+    // Events
+    event Mint(address _soul);
+    event Burn(address _soul);
+    event Update(address _soul);
+
+    /**
+     * @dev This modifier checks that the address passed in is not the zero address.
+     */
+    modifier validAddress(address _addr) {
+        require(_addr != address(0), "Not valid address");
+        _;
+    }
+
+    /**
+     * @dev Initializes the contract by setting a `name` and a `symbol` to the SBT
+     */
+    constructor(string memory name_, string memory symbol_) {
+        _name = name_;
+        _symbol = symbol_;
+        _totalSBT = 0;
+    }
+
+    /**
+     * @dev Mints `SBT` and transfers it to `_soul`.
+     *
+     * Emits a {Mint} event.
+     */
+    function mint(address _soul, Soul memory _soulData)
+        external
+        virtual
+        validAddress(_soul)
+    {
+        require(!hasSoul(_soul), "Soul already exists");
```

Comment:
This assumes that for each contract, each address can only have one soul - is that the intention of our implementation?

Reply:
yup i was thinking that if Aave were to deploy a SBT, each address (Soul) can only have one SBT from aave. But each Soul can have multiple SBT from different projects.

Reply:
Ok gotcha on that - this concept should be in the article if it's not yet in

Reply:
got it ~

Reply:
Do add some examples